File: wliotproxy-src/libwliotproxy/IotServerCommandCall.cpp
#include "IotServerCommandCall.h"
#include "IotServerConnection.h"
#include "wliot/WLIOTProtocolDefs.h"
#include "wliot/WLIOTServerProtocolDefs.h"
#include <QCoreApplication>
#include <QDebug>

IotServerCommandCall::IotServerCommandCall(IotServerConnection *conn,CmDataCallback onCmData,
	const QByteArray &cmd,const QByteArrayList &args,QObject *parent)
	:QObject(parent)
{
	srvConn=conn;
	mCmd=cmd;
	mArgs=args;
	mOnCmData=onCmData;
	static qint64 callIdNum=0;
	callId=QByteArray::number(++callIdNum);
	done=false;
	mOk=false;

	tmr.setSingleShot(true);
	tmr.setInterval(WLIOTProtocolDefs::syncWaitTime*4);

	//CRIT think about timers for a server
	connect(&tmr,&QTimer::timeout,this,&IotServerCommandCall::onTimeout);
	connect(conn,&IotServerConnection::disconnected,this,&IotServerCommandCall::onDisconnected,Qt::QueuedConnection);
	connect(conn,&IotServerConnection::funcCallReplyMsg,this,&IotServerCommandCall::onMessage,Qt::QueuedConnection);
}

void IotServerCommandCall::call()
{
	if(!srvConn->isConnected())
	{
		mOk=false;
		done=true;
		retVal.append("disconnected");
		return;
	}
	tmr.start();
	qDebug()<<"Command to server: "<<mCmd<<":"<<callId<<":"<<mArgs;
	srvConn->writeMsg(Message(mCmd,QByteArrayList()<<callId<<mArgs));
	while(!done)
		qApp->processEvents(QEventLoop::WaitForMoreEvents);
	if(mOk)
		qDebug()<<"Command ok: "<<callId<<":"<<retVal;
	else qDebug()<<"Command err: "<<callId<<":"<<retVal;
}

bool IotServerCommandCall::ok()
{
	return mOk;
}

const QByteArrayList &IotServerCommandCall::returnValue()
{
	return retVal;
}

void IotServerCommandCall::onMessage(const Message &m)
{
	if(m.args.isEmpty())return;
	if(m.title==WLIOTProtocolDefs::funcAnswerOkMsg)
	{
		if(m.args[0]!=callId)return;
		QMetaObject::invokeMethod(this,"setDone",Qt::QueuedConnection);
		mOk=true;
		retVal=m.args.mid(1);
		tmr.stop();
	}
	else if(m.title==WLIOTProtocolDefs::funcAnswerErrMsg)
	{
		if(m.args[0]!=callId)return;
		QMetaObject::invokeMethod(this,"setDone",Qt::QueuedConnection);
		retVal=m.args.mid(1);
		tmr.stop();
	}
	else if(m.title==WLIOTServerProtocolDefs::srvCmdDataMsg)
	{
		if(m.args[0]!=callId||!mOnCmData)return;
		if(!mOnCmData(m.args.mid(1)))
		{
			QMetaObject::invokeMethod(this,"setDone",Qt::QueuedConnection);
			tmr.stop();
		}
	}
}

void IotServerCommandCall::onDisconnected()
{
	QMetaObject::invokeMethod(this,"setDone",Qt::QueuedConnection);
	retVal.append("server disconnected");
}

void IotServerCommandCall::onTimeout()
{
	if(done)return;
	QMetaObject::invokeMethod(this,"setDone",Qt::QueuedConnection);
	retVal.append("timeout");
}

void IotServerCommandCall::setDone()
{
	done=true;
}

File: WLIOTAndroidApplication/libwliotproxy/src/main/java/com/alterozoom/libwliotproxy/IotServerCommandCall.java
package com.alterozoom.libwliotproxy;

import java.util.Timer;
import java.util.TimerTask;

public class IotServerCommandCall
{
	private class ConnEvCb
		extends IotServerConnection.EventsCallback
	{
		void onDisconnected()
		{
			retVal.add(ByteArray.fromString("server disconnected"));
			waitObj.notify();
		}

		void funcCallReplyMsg(Message m)
		{
			if(m.args.size()==0)return;
			if(m.title.equals(WLIOTProtocolDefs.funcAnswerOkMsg))
			{
				if(!m.args.get(0).equals(callId))return;
				mOk=true;
				retVal=new ByteArrayList(m.args);
				retVal.remove(0);
				waitObj.notify();
			}
			else if(m.title.equals(WLIOTProtocolDefs.funcAnswerErrMsg))
			{
				if(!m.args.get(0).equals(callId))return;
				retVal=new ByteArrayList(m.args);
				retVal.remove(0);
				waitObj.notify();
			}
			else if(m.title.equals(WLIOTServerProtocolDefs.srvCmdDataMsg))
			{
				if(!m.args.get(0).equals(callId)||mOnCmData==null)return;
				ByteArrayList lCopy=new ByteArrayList(m.args);
				lCopy.remove(0);
				if(!mOnCmData.onCmData(lCopy))
					waitObj.notify();
			}
		}
	}

	private ConnEvCb evCb;
	private IotServerConnection srvConn;
	private CmDataCallback mOnCmData;
	private ByteArray mCmd;
	private ByteArrayList mArgs;
	private ByteArrayList retVal;
	private ByteArray callId;
	private boolean mOk;
	private static long callIdNum=0;
	private Object waitObj;
	private Timer tmr;

	public IotServerCommandCall(IotServerConnection conn,CmDataCallback onCmData,final ByteArray cmd,
		final ByteArrayList args)
	{
		srvConn=conn;
		mOnCmData=onCmData;
		mCmd=cmd;
		mArgs=args;
		evCb=new ConnEvCb();
		srvConn.registerCallback(evCb);
		callId=ByteArray.fromString(Long.toString(++callIdNum));
		tmr=new Timer();
		waitObj=new Object();
		mOk=false;
	}

	public void finalize()
	{
		srvConn.removeCallback(evCb);
	}

	public void call()
	{
		if(!srvConn.isConnected())
		{
			mOk=false;
			retVal.add(ByteArray.fromString("disconnected"));
			return;
		}
		if(mArgs==null)
			srvConn.writeMsg(new Message(mCmd,new ByteArrayListBuilder().add(callId).done()));
		else srvConn.writeMsg(new Message(mCmd,new ByteArrayListBuilder().add(callId).add(mArgs).done()));
		tmr.schedule(new TimerTask(){
			@Override
			public void run()
			{
				retVal.add(ByteArray.fromString("timeout"));
				mOk=false;
				waitObj.notify();
			}
		},WLIOTProtocolDefs.syncWaitTime*4);
		try
		{
			waitObj.wait();
		}
		catch(InterruptedException e)
		{
			e.printStackTrace();
			mOk=false;
			retVal.add(ByteArray.fromString("exception catched"));
		}
	}

	boolean ok()
	{
		return mOk;
	}

	ByteArrayList returnValue()
	{
		return retVal;
	}
}
